### DISCUSSION TITLE
Allow injecting SQLite extensions

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#7758] [Feature Requests]
SQLite extensions—unlike any other DB—are loaded on the connection instead of the "database" level. This means that in order to use any SQLite extension, we have to allow for an environment variable to set the paths to the `dylib` files for the extensions, and use [`.loadExtension()` from SQLite](https://github.com/mapbox/node-sqlite3/wiki/Extensions) to pull them in right when the connection is created.

We should be able to rely on Node's "afterCreate" pool configuration for this, the following basic example illustrates this idea:

```js
const knex = require("knex");

const db = knex({
	client: "sqlite",
	connection: "./test.db",
	pool: {
		afterCreate(conn, callback) {
			conn.loadExtension("/usr/local/lib/mod_spatialite.dylib", (err) => {
				callback(err, conn);
			});
		},
	},
});

db.raw(`SELECT spatialite_version();`)
	.then((res) => console.log(res))
	.catch((res) => console.error(res));
```

(this example seems to work as expected)

However, when we apply the same idea to the injection point in Directus:

https://github.com/directus/directus/blob/45b8c12dee649fd9a80b50a5451b1f524d04a299/api/src/database/index.ts#L72-L76

As follows

```js
poolConfig.afterCreate = (conn: any, cb: any) => { 
  conn.loadExtension("/usr/local/lib/mod_spatialite.dylib", cb);
}; 
```

the process crashes when a segmentation fault as soon as the database is used (like in `validateDBConnection()`) in app.ts. 

At this point, I don't know what's causing that segmentation error. It could be the total memory used in the directus process itself (i'm assuming node-sqlite3 reads the extension into memory for it to work), it could be that spatialite itself access illegal memory, or a range of other reasons. 

---

Here's a more in-depth implementation for full dynamic extensions support:

```js
poolConfig.afterCreate = async (conn: any, callback: any) => {
	logger.trace('Enabling SQLite Foreign Keys support...');

	await run('PRAGMA foreign_keys = ON');

	for (const extensionPath of toArray(env.DB_EXTENSIONS ?? [])) {
		logger.trace(`Loading extension "${extensionPath}"...`);
		await loadExtension(extensionPath);
	}

	callback(null, conn);

	async function run(sql: string): Promise<any> {
		return new Promise((resolve, reject) => {
			conn.run(sql, (err: Error | null, result: any) => {
				if (err) return reject(err);
				resolve(result);
			});
		});
	}

	async function loadExtension(extensionPath: string): Promise<any> {
		return new Promise((resolve, reject) => {
			conn.loadExtension(extensionPath, (err: Error | null, result: any) => {
				if (err) return reject(err);
				return resolve(result);
			});
		});
	}
};
```

---

In general, it turns out that SQLite extensions are pretty finicky, relying on system-bound compiled c-files (most of which I can't get to compile). `node-sqlite3` will attempt to use the system installed SQLite version first, and falls back to a included dist second. If the main system sqlite instance already has extensions statically linked, things should work as expected without the reliance on injecting them through Directus. If this problem persists and doesn't have a cause we can resolve within Directus itself, this feature might simply not be possible.

**Comment by winterxc** [comment]
current solution:
```
#define a hook
module.exports=({ init}, {database})=>{
    const pool = database.client.pool
    pool.acquire().promise.then(async (conn)=>{
        await conn.loadExtension('/usr/local/lib/libsimple')
        pool.release(conn)
        console.log('load extension done')
    })
    init('app.after', ({app}) => {
        //todo
    })
}
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
